components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,395 @@
+              return std::tie(first.block_number, first.log_index) <
+                     std::tie(second.block_number, second.log_index);
+            });
+
+  // Collection of the latest allowances per contract & spender & approver.
+  std::map<std::string, mojom::AllowanceInfoPtr> allowance_map;
+  // Put cached data into the map first if available.
+  LoadCachedAllowances(task_iter->second->chain_id_,
+                       task_iter->second->account_address_, allowance_map);
+  for (const auto& log_item : logs_tmp) {
+    // Skip pending logs.
+    if (log_item.block_number == 0) {
+      continue;
+    }
+
+    if (log_item.topics.size() < kTopicElementsCountToCheck) {
+      continue;
+    }
+    const auto current_map_key = GetAllowanceMapKey(
+        log_item.address, log_item.topics[1], log_item.topics[2]);
+
+    uint256_t parsed_amount{0};
+    if (!HexValueToUint256(log_item.data, &parsed_amount)) {
+      continue;
+    }
+
+    if (parsed_amount > 0) {
+      // Replace if same key exists by the fresh allowance data.
+      allowance_map.insert_or_assign(
+          current_map_key,
+          mojom::AllowanceInfo::New(task_iter->second->chain_id_,
+                                    log_item.address, log_item.topics[1],
+                                    log_item.topics[2], log_item.data));
+    } else if (allowance_map.find(current_map_key) != allowance_map.end()) {
+      allowance_map.erase(current_map_key);
+    }
+  }
+  std::vector<mojom::AllowanceInfoPtr> allowances_found;
+  for (const auto& [allowance_map_key, allowance] : allowance_map) {
+    allowances_found.push_back(allowance.Clone());
+  }
+  task_iter->second->SetResults(std::move(allowances_found));
+  MaybeMergeAllResultsAndCallBack();
+}
+
+bool EthAllowanceManager::IsAllTasksCompleted() const {
+  DCHECK(!discover_eth_allowance_callbacks_.empty());
+
+  return get_block_tasks_ == 0 &&
+         base::ranges::all_of(allowance_discovery_tasks_, [](const auto& item) {
+           return item.second->is_completed_;
+         });
+}
+
+void EthAllowanceManager::MaybeMergeAllResultsAndCallBack() {
+  if (!IsAllTasksCompleted()) {
+    return;
+  }
+  std::vector<mojom::AllowanceInfoPtr> result;
+  ScopedDictPrefUpdate allowance_cache_update(prefs_,
+                                              kBraveWalletEthAllowancesCache);
+  auto& allowance_cache = allowance_cache_update.Get();
+
+  std::map<std::string, std::vector<std::unique_ptr<EthAllowanceTask>>>
+      allowance_tasks_by_chain_ids;
+  for (auto& [tid, allowance_task_info] : allowance_discovery_tasks_) {
+    if (!allowance_tasks_by_chain_ids.contains(
+            allowance_task_info->chain_id_)) {
+      allowance_tasks_by_chain_ids.insert_or_assign(
+          allowance_task_info->chain_id_,
+          std::vector<std::unique_ptr<EthAllowanceTask>>());
+    }
+    allowance_tasks_by_chain_ids[allowance_task_info->chain_id_].push_back(
+        std::move(allowance_task_info));
```

Comment:
This one creates empty vector value if no chain_id_ key in map
```
    allowance_tasks_by_chain_ids[allowance_task_info->chain_id_].push_back(
        std::move(allowance_task_info));
```
so this one is not needed
```
    if (!allowance_tasks_by_chain_ids.contains(
            allowance_task_info->chain_id_)) {
      allowance_tasks_by_chain_ids.insert_or_assign(
          allowance_task_info->chain_id_,
          std::vector<std::unique_ptr<EthAllowanceTask>>());
    }
```

Reply:
fixed